src/main/java/com/google/devtools/build/remote/client/AbstractRemoteActionCache.java

```diff
@@ -15,17 +15,183 @@
 package com.google.devtools.build.remote.client;
 
 import com.google.devtools.remoteexecution.v1test.Digest;
+import com.google.devtools.remoteexecution.v1test.Directory;
+import com.google.devtools.remoteexecution.v1test.DirectoryNode;
+import com.google.devtools.remoteexecution.v1test.FileNode;
+import com.google.devtools.remoteexecution.v1test.Tree;
+import com.google.protobuf.ByteString;
 import java.io.IOException;
 import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.PosixFilePermission;
+import java.util.ArrayList;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.Nullable;
 
 /** A cache for storing artifacts (input and output) as well as the output of running an action. */
 public abstract class AbstractRemoteActionCache {
+
+  public DigestUtil getDigestUtil() {
+    return digestUtil;
+  }
+
   protected final DigestUtil digestUtil;
 
   public AbstractRemoteActionCache(DigestUtil digestUtil) {
     this.digestUtil = digestUtil;
   }
 
-  /** Downloads a single blob with the specified digest into an output stream. */
+  /**
+   * Download a tree with the {@link Directory} given by digest as the root directory of the tree.
+   *
+   * @param rootDigest The digest of the root {@link Directory} of the tree
+   * @return A tree with the given directory as the root.
+   * @throws IOException in the case that retrieving the blobs required to reconstruct the tree
+   *     failed.
+   */
+  public Tree getTree(Digest rootDigest) throws IOException {
+    byte directoryBytes[];
+    Directory dir;
+    directoryBytes = downloadBlob(rootDigest);
+    dir = Directory.parseFrom(directoryBytes);
+    return getTree(dir);
+  }
+
+  // A default implementation for recursively downloading the tree associated with a directory.
+  // Note: This implementation requires a round trip for each directory in the directory tree.
+  protected Tree getTree(Directory rootDir) throws IOException {
+    List<Directory> children = new ArrayList<>();
+    addChildDirectories(children, rootDir);
+    return Tree.newBuilder().setRoot(rootDir).addAllChildren(children).build();
+  }
+
+  // Recursively add all child directories to the given list.
+  private void addChildDirectories(List<Directory> directories, Directory dir) throws IOException {
```

Comment:
Nit: could you move the output parameter to the end?  I think the convention would be to have input parameters precede output.

In general the [Java Practices](https://engdoc.corp.google.com/eng/doc/devguide/java/practices/methods.md?cl=head)  suggests avoiding output parameters where possible, though I agree that in this case it might make sense to have it.

Reply:
done. I wasn't too familiar with whether repeatedly concatenating the lists for a more functional solution would have poor performance, so I went with the safer bet of accumulating the directories in an output parameter